### PR TITLE
refactored listener function

### PR DESCRIPTION
**Pull request by shark8me**
A couple of changes to the listener function which is run on every test epoch:

- Reduces the number of arguments, now has 1 map for global context and another for per-epoch context
- train dataset is also passed to the listener function to evaluate metrics on training set.

I felt that saving the network is best implemented by the listener function, since the requirements may differ. In some situations, each network iteration needs to be saved and in others, the best current network needs to be saved.
Refactored the code to save the network within the listener function.

### CODE REVIEW

**Comment by harold** [comment]
Thanks for this!

I had some ideas for similar changes, so I'll consider starting with these when I come back to this area.

**Comment by shark8me** [comment]
I want to add that this change enables listeners (such as the example in #172 ) to compare training set and test/validation set performance, which isn't possible at the moment.

File: experiment/src/cortex/experiment/classification.clj [hunk 1]
```diff
@@ -188,18 +192,26 @@
...
+                          (let [best-network
+                            (assoc new-network 
+                              :classification-accuracy classification-accuracy)]
+                            (reset-confusion-matrix confusion-matrix-atom
+                                                    observation->img-fn
+                                                    class-mapping
+                                                    {:labels labels
+                                                    :test-ds test-ds})
+                            (experiment-train/save-network best-network network-filename))
+                            ;;seems dicey. if not the best-network,
```
Comment (harold) [inline experiment/src/cortex/experiment/classification.clj]: You're totally right.

For this simple example, this performs well. But moving forward we'll be exposing more of our sophisticated training regimes as appropriate.

**Comment by harold** [comment]
Moved the branch to our repo: https://github.com/thinktopic/cortex/pull/175
